integration-tests/src/tests/client/features/yield_timeouts.rs

```diff
@@ -0,0 +1,366 @@
+    let last_block_height = env.clients[0].chain.head().unwrap().height;
+
+    let mut result = vec![];
+
+    for receipt in env.clients[0]
+        .chain
+        .get_outgoing_receipts_for_shard(last_block_hash, shard_id, last_block_height)
+        .unwrap()
+    {
+        if let PromiseYield(ref action_receipt) = receipt.receipt {
+            result.push(action_receipt.input_data_ids[0]);
+        }
+        if let PromiseResume(ref data_receipt) = receipt.receipt {
+            result.push(data_receipt.data_id);
+        }
+    }
+
+    result
+}
+
+/// Create environment with an unresolved promise yield callback.
+/// Returns the test environment, the yield tx hash, and the data id for resuming the yield.
+fn prepare_env_with_yield(
+    anticipated_yield_payload: Vec<u8>,
+    test_env_gas_limit: Option<u64>,
+) -> (TestEnv, CryptoHash, CryptoHash) {
+    init_test_logger();
+    let mut genesis = Genesis::test(vec!["test0".parse().unwrap(), "test1".parse().unwrap()], 1);
+    if let Some(gas_limit) = test_env_gas_limit {
+        genesis.config.gas_limit = gas_limit;
+    }
+    let mut env = TestEnv::builder(&genesis.config).nightshade_runtimes(&genesis).build();
+    let genesis_block = env.clients[0].chain.get_block_by_height(0).unwrap();
+    let signer = InMemorySigner::from_seed("test0".parse().unwrap(), KeyType::ED25519, "test0");
+
+    // Submit transaction deploying contract to test0
+    let tx = SignedTransaction::from_actions(
+        1,
+        "test0".parse().unwrap(),
+        "test0".parse().unwrap(),
+        &signer,
+        vec![Action::DeployContract(DeployContractAction {
+            code: near_test_contracts::nightly_rs_contract().to_vec(),
+        })],
+        *genesis_block.hash(),
+    );
+    let tx_hash = tx.get_hash();
+    assert_eq!(env.clients[0].process_tx(tx, false, false), ProcessTxResponse::ValidTx);
+
+    // Allow two blocks for the contract to be deployed
+    for i in 1..3 {
+        env.produce_block(0, i);
+    }
+    assert!(matches!(
+        env.clients[0].chain.get_final_transaction_result(&tx_hash).unwrap().status,
+        FinalExecutionStatus::SuccessValue(_),
+    ));
+
+    // Submit transaction making a function call which will invoke yield create
+    let yield_transaction = SignedTransaction::from_actions(
+        10,
+        "test0".parse().unwrap(),
+        "test0".parse().unwrap(),
+        &signer,
+        vec![Action::FunctionCall(Box::new(FunctionCallAction {
+            method_name: "call_yield_create_return_promise".to_string(),
+            args: anticipated_yield_payload,
+            gas: 300_000_000_000_000,
+            deposit: 0,
+        }))],
+        *genesis_block.hash(),
+    );
+    let yield_tx_hash = yield_transaction.get_hash();
+    assert_eq!(
+        env.clients[0].process_tx(yield_transaction, false, false),
+        ProcessTxResponse::ValidTx
+    );
+
+    // Allow two blocks for the function call to be executed
+    for i in 3..5 {
```

Comment:
```suggestion
    for i in 3..NEXT_BLOCK_HEIGHT_AFTER_SETUP {
```

perhaps? I would also introduce a variable for the `3`. An assert that the variabl does not exceed the constant wouldn't hurt either.

Reply:
I simply added a check at the end of `prepare_env` which gets the height of the head of the chain from one of the test clients and compares with `NEXT_BLOCK_HEIGHT_AFTER_SETUP`.

I think changing these constants to variables doesn't achieve much:
- It obfuscates what the loop wants to do (advance two blocks)
- The height `3` is never referenced outside the `prepare_env` function